internal/observerware/call.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package observerware
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+// A call represents a single RPC along an edge.
+//
+// To prevent allocating on the heap on the request path, it's a value instead
+// of a pointer.
+type call struct {
+	edge    *edge
+	extract ContextExtractor
+
+	started time.Time
+	ctx     context.Context
+	req     *transport.Request
+	rpcType transport.Type
+	inbound bool
+}
+
+func (c call) End(err error, isApplicationError bool) {
+	elapsed := _timeNow().Sub(c.started)
+	c.endLogs(elapsed, err, isApplicationError)
+	c.endStats(elapsed, err, isApplicationError)
+}
+
+func (c call) endLogs(elapsed time.Duration, err error, isApplicationError bool) {
+	msg := "Handled inbound request."
+	if !c.inbound {
+		msg = "Made outbound call."
+	}
+	if ce := c.edge.logger.Check(zap.DebugLevel, msg); ce != nil {
+		fields := make([]zapcore.Field, 0, 5)
+		fields = append(fields, zap.String("rpcType", c.rpcType.String()))
```

Comment:
It doesn't make it slower, but since the strings are only used once in the production code I don't see the gain in readability.